src/datasets/arrow_dataset.py

```diff
@@ -1601,6 +1596,16 @@ def apply_function_on_filtered_inputs(inputs, indices, check_same_num_examples=F
                     for i, example in enumerate(pbar):
                         example = apply_function_on_filtered_inputs(example, i, offset=offset)
                         if update_data:
+                            if i == 0:
+                                cache_file_name = determine_cache_file_name(cache_file_name)
+                                if check_if_cached(cache_file_name):
+                                    logger.warning("Loading cached processed dataset at %s", cache_file_name)
+                                    info = self.info.copy()
+                                    info.features = features
+                                    return Dataset.from_file(cache_file_name, info=info, split=self.split)
```

Comment:
This can be outside the for loop, no ?
I don't think we need the first example to know if we can reload it from the cache

Reply:
We need the value of `update_data`, which we know after processing the first example, to determine the name of the cache file (and to reload the dataset if this file already exists), so we have to enter the loop. 

The https://github.com/huggingface/datasets/pull/1960/commits/c38110c9b5a22e645eadd1a766b4fc270bd23758 commit has the approach where the "return" part is outside the loop, but I think the current solution is cleaner and easier to follow.

Reply:
I don't think we need the value of `update_data` to get the cache file name ?
Before the loop you can directly compute the cache file name and if it exists and load_from_cache_file is True, then reload the dataset before entering the loop.
Does that make sense or I missed something ?